Create CI strategy (#31)
buerokratt/Cross-functional-requirements: docs/technical/continuous-integration-strategy.md

Continuous Integration (CI) is the practice of merging developer's code to the central (`main`) branch frequently and in doing so "continuously" ensuring that the code "integrates" well with other code. 

It is typicaly when practicing CI, that the incoming code (the code being merged) goes through several automated checks before merging, this is the purpose of the CI pipeline.

The code for Bürokratt  is written in C# which mandates some of the things that the CI pipeline does.

## When does CI run?

CI pipelines should run as part of the PR process when merging features branches to `main`. A PR should have a "green" CI pipeline before code can be merged and the PR can be completed, this means that all automated checks in the CI pipeline must pass.

## What does CI do?

All CI pipelines will do the following tasks

| Task             | Description                                                  |
| ---------------- | ------------------------------------------------------------ |
| dotnet format    | Uses the `dotnet` CLI to format the code according to defined style rules for the project. See https://github.com/dotnet/format |

> | dotnet restore   | Uses the `dotnet` CLI to restore the .net packages which the project depends on. See https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-restore |

Reviewer: ```suggestion
| dotnet restore   | Uses the `dotnet` CLI to restore the .NET packages which the project depends on. See https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-restore |
```

> | dotnet build     | Uses the `dotnet` CII to build the project. See https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-build |

Reviewer: ```suggestion
| dotnet build     | Uses the `dotnet` CLI to build the project. See https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-build |
```

> | dotnet test      | Uses the `dotnet` CLI to run all tests associated with the project. This include integration and unit tests. See https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-test |

Reviewer: ```suggestion
| dotnet test      | Uses the `dotnet` CLI to run all tests associated with the project. This includes integration and unit tests. See https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-test |
```

| Mutation testing | Runs the mutation testing package as defined in [unit-testing-strategy](unit-testing-strategy.md). |

> | dotnet publish   | Uses the `dotnet` CI to publish the project. This makes the files available for packaging. |

Reviewer: ```suggestion
| dotnet publish   | Uses the `dotnet` CLI to publish the project. This produces deployable binaries ready for deployment. |
```

| Build image      | Uses the `docker` CLI to build the image for the project     |
| Sign image       | Signs the docker image for the project                       |

> | Upload image     | Uploads the Docker image to the GitHub container registry    |

Reviewer: Just using consistent wording for uploading docker images (i.e. - 'push')
```suggestion
| Push image     | Pushes the Docker image to the GitHub container registry    |
```
